lxd/device/unix_hotplug.go

```diff
@@ -0,0 +1,215 @@
+			if err != nil {
+				return nil, err
+			}
+
+			// Add a post hook function to remove the specific unix hotplug device file after unmount.
+			runConf.PostHooks = []func() error{func() error {
+				err := unixDeviceDeleteFiles(state, devicesPath, "unix", deviceName, relativeTargetPath)
+				if err != nil {
+					return fmt.Errorf("Failed to delete files for device '%s': %v", deviceName, err)
+				}
+
+				return nil
+			}}
+		}
+
+		runConf.Uevents = append(runConf.Uevents, e.UeventParts)
+
+		return &runConf, nil
+	}
+
+	unixHotplugRegisterHandler(d.instance, d.name, f)
+
+	return nil
+}
+
+// Start is run when the device is added to the instance
+func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
+	runConf := deviceConfig.RunConfig{}
+	runConf.PostHooks = []func() error{d.Register}
+
+	device := d.loadUnixDevice()
+	if d.isRequired() && device == nil {
+		return nil, fmt.Errorf("Required Unix Hotplug device not found")
+	}
+	if device == nil {
+		return &runConf, nil
+	}
+
+	i, err := strconv.ParseUint(device.PropertyValue("MAJOR"), 10, 32)
+	if err != nil {
+		return nil, err
+	}
+	major := uint32(i)
+	j, err := strconv.ParseUint(device.PropertyValue("MINOR"), 10, 32)
+	if err != nil {
+		return nil, err
+	}
+	minor := uint32(j)
+
+	// setup device
+	if device.Subsystem() == "char" {
+		err = unixDeviceSetupCharNum(d.state, d.instance.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+	} else if device.Subsystem() == "block" {
+		err = unixDeviceSetupBlockNum(d.state, d.instance.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &runConf, nil
+}
+
+// Stop is run when the device is removed from the instance
+func (d *unixHotplug) Stop() (*deviceConfig.RunConfig, error) {
+	unixHotplugUnregisterHandler(d.instance, d.name)
+
+	runConf := deviceConfig.RunConfig{
+		PostHooks: []func() error{d.postStop},
+	}
+
+	err := unixDeviceRemove(d.instance.DevicesPath(), "unix", d.name, "", &runConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &runConf, nil
+}
+
+// postStop is run after the device is removed from the instance
+func (d *unixHotplug) postStop() error {
+	err := unixDeviceDeleteFiles(d.state, d.instance.DevicesPath(), "unix", d.name, "")
+	if err != nil {
+		return fmt.Errorf("Failed to delete files for device '%s': %v", d.name, err)
+	}
+
+	return nil
+}
+
+// loadUnixDevice scans the host machine for unix devices with matching product/vendor ids
+// and returns the first matching device with the subsystem type char or block
+func (d *unixHotplug) loadUnixDevice() *udev.Device {
+	// Find device if exists
+	u := udev.Udev{}
+	e := u.NewEnumerate()
+
+	if d.config["vendorid"] != "" {
+		e.AddMatchProperty("ID_VENDOR_ID", d.config["vendorid"])
+	}
+	if d.config["productid"] != "" {
+		e.AddMatchProperty("ID_MODEL_ID", d.config["productid"])
```

Comment:
Just so we're clear: both, `ID_VENDOR_ID` and `ID_MODEL_ID` are properties added by udev itself and not the kernel. The kernel fields are `idVendor` and `idProduct` which are parsed by udev which the sets up `ID_{VENDOR,MODEL}_ID` accordingly.
Depending on whether you parse `udev` uevent output or pure `kernel` uevent output this matters.